client/blocks/qr-code-login/index.jsx

```diff
@@ -0,0 +1,198 @@
+
+import './style.scss';
+
+const AUTH_PULL_INTERVAL = 5000; // 5 seconds
+const LOCALE_STORAGE_KEY = 'qr-login-token';
+
+const isStillValidToken = ( tokenData ) => {
+	if ( ! tokenData ) {
+		return false;
+	}
+	const { expires } = tokenData;
+	if ( ! expires ) {
+		return false;
+	}
+	return expires > Date.now() / 1000;
+};
+
+const setLocalTokenData = ( data ) => {
+	try {
+		window.localStorage.setItem( LOCALE_STORAGE_KEY, JSON.stringify( data ) );
+	} catch ( e ) {}
+};
+
+const getLocalTokenData = () => {
+	try {
+		const valueString = window.localStorage.getItem( LOCALE_STORAGE_KEY );
+		if ( valueString === undefined || valueString === null ) {
+			return false;
+		}
+
+		const tokenData = JSON.parse( valueString );
+		if ( isStillValidToken( tokenData ) ) {
+			return tokenData;
+		}
+	} catch ( e ) {}
+
+	return false;
+};
+
+const getLoginActionResponse = async ( action, args ) => {
+	const url = new URL( 'https://wordpress.com/wp-login.php' );
+	url.searchParams.append( 'action', action );
+
+	Object.keys( args ).forEach( ( key ) => {
+		url.searchParams.append( key, args[ key ] );
+	} );
+
+	const response = await fetch( url.href );
+	return await response.json();
+};
+
+const fetchQRCodeData = async ( setTokenData, tokenData, anonymousUserId ) => {
+	if ( isStillValidToken( tokenData ) ) {
+		return tokenData;
+	}
+
+	const responseData = await getLoginActionResponse( 'qr-code-token-request-endpoint', {
+		anon_id: anonymousUserId,
+	} );
+
+	setTokenData( responseData.data );
+	setLocalTokenData( responseData.data );
+};
+
+const fetchAuthState = async ( setAuthState, setTokenData, tokenData, anonymousUserId ) => {
+	if ( ! tokenData ) {
+		return;
+	}
+
+	if ( ! isStillValidToken( tokenData ) ) {
+		fetchQRCodeData( setTokenData, tokenData, anonymousUserId );
+		return;
+	}
+
+	const { token, encrypted } = tokenData;
+
+	const responseData = await getLoginActionResponse( 'qr-code-authentication-endpoint', {
+		anon_id: anonymousUserId,
+		token: token,
+		data: encrypted,
+	} );
+
+	setAuthState( responseData.data );
+};
+
+function TokenQRCode( { tokenData } ) {
+	if ( ! tokenData ) {
+		return <QRCodePlaceholder />;
+	}
+	const { token, encrypted } = tokenData;
+	const imageSettings = {
+		src: qrCenter,
+		height: 64,
+		width: 64,
+		excavate: false,
+	};
+	return (
+		<QRCode
+			value={ `https://apps.wordpress.com/get/?campaign=login-qr-code#qr-code-login?token=${ token }&data=${ encrypted }` }
+			size={ 352 }
+			imageSettings={ imageSettings }
+		/>
+	);
+}
+
+function QRCodePlaceholder() {
+	return (
+		<div className="qr-code-login__placeholder">
+			<span className="qr-code-login__corner-box"></span>
+			<span className="qr-code-login__corner-box"></span>
+			<span className="qr-code-login__corner-box"></span>
+		</div>
+	);
+}
+
+function QRCodeLogin() {
+	const translate = useTranslate();
+	const [ tokenData, setTokenData ] = useState( getLocalTokenData() );
```

Comment:
`getLocalTokenData()` will be called on every render, regardless of the fact that the value will be used just for setting the initial state. Reading from local storage can sometimes be slow, so it's recommended that this doesn't run needlessly.

To prevent from doing this, you can use the callback form to lazy initialize the state:

```suggestion
	const [ tokenData, setTokenData ] = useState( () => getLocalTokenData() );
```

You can read more about that here: https://reactjs.org/docs/hooks-reference.html#lazy-initial-state